File: src/portfolio.zig
const std = @import("std");
const api = @import("api.zig");

pub const Item = api.PortfolioItem;

pub const Config = struct {
    portfolio: []Item = &[_]Item{},
    refresh_interval: i64 = 5,

    pub fn deinit(self: *Config, allocator: std.mem.Allocator) void {
        for (self.portfolio) |item| {
            allocator.free(item.market);
            allocator.free(item.code);
            allocator.free(item.name);
        }
        allocator.free(self.portfolio);
    }
};

fn getConfigPath(allocator: std.mem.Allocator) ![]u8 {
    const home = std.process.getEnvVarOwned(allocator, "HOME") catch {
        return allocator.dupe(u8, ".leek_tui.json");
    };
    defer allocator.free(home);
    return try std.fs.path.join(allocator, &.{ home, ".leek_tui.json" });
}

fn migrateFromOld(allocator: std.mem.Allocator) !?[]Item {
    const home = std.process.getEnvVarOwned(allocator, "HOME") catch return null;
    defer allocator.free(home);

    const old_path = try std.fs.path.join(allocator, &.{ home, ".config", "leek-tui", "portfolio.json" });
    defer allocator.free(old_path);

    const data = std.fs.cwd().readFileAlloc(allocator, old_path, 1024 * 1024) catch |e| {
        if (e == error.FileNotFound) return null;
        return e;
    };
    defer allocator.free(data);

    const parsed = try std.json.parseFromSlice([]Item, allocator, data, .{});
    defer parsed.deinit();

    var items = try allocator.alloc(Item, parsed.value.len);
    for (parsed.value, 0..) |src, i| {
        items[i] = .{
            .market = try allocator.dupe(u8, src.market),
            .code = try allocator.dupe(u8, src.code),
            .name = try allocator.dupe(u8, src.name),
        };
    }
    return items;
}

pub fn load(allocator: std.mem.Allocator) !Config {
    const path = try getConfigPath(allocator);
    defer allocator.free(path);

    const data = std.fs.cwd().readFileAlloc(allocator, path, 1024 * 1024) catch |e| {
        if (e == error.FileNotFound) {
            if (try migrateFromOld(allocator)) |items| {
                return Config{ .portfolio = items };
            }
            return Config{};
        }
        return e;
    };
    defer allocator.free(data);

    const parsed = try std.json.parseFromSlice(Config, allocator, data, .{});
    defer parsed.deinit();

    // Deep copy portfolio since parsed will be deinit
    var items = try allocator.alloc(Item, parsed.value.portfolio.len);
    for (parsed.value.portfolio, 0..) |src, i| {
        items[i] = .{
            .market = try allocator.dupe(u8, src.market),
            .code = try allocator.dupe(u8, src.code),
            .name = try allocator.dupe(u8, src.name),
        };
    }

    return Config{
        .portfolio = items,
        .refresh_interval = parsed.value.refresh_interval,
    };
}

pub fn save(allocator: std.mem.Allocator, config: Config) !void {
    const path = try getConfigPath(allocator);
    defer allocator.free(path);

    const file = try std.fs.cwd().createFile(path, .{});
    defer file.close();

    const json_bytes = try std.json.Stringify.valueAlloc(allocator, config, .{ .whitespace = .indent_2 });
    defer allocator.free(json_bytes);
    try file.writeAll(json_bytes);
    try file.writeAll("\n");
}

File: src/portfolio.ts
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { homedir } from 'node:os';
import { join } from 'node:path';
import type { Config, PortfolioItem } from './types.js';

const CONFIG_NAME = '.leek_tui.json';
const OLD_CONFIG_DIR = join(homedir(), '.config', 'leek-tui');
const OLD_CONFIG_PATH = join(OLD_CONFIG_DIR, 'portfolio.json');

function getConfigPath(): string {
  return join(homedir(), CONFIG_NAME);
}

async function migrateFromOld(): Promise<PortfolioItem[] | null> {
  try {
    const data = await readFile(OLD_CONFIG_PATH, 'utf-8');
    const parsed = JSON.parse(data) as PortfolioItem[];
    return parsed.map((item) => ({
      market: String(item.market),
      code: String(item.code),
      name: String(item.name),
    }));
  } catch {
    return null;
  }
}

export async function load(): Promise<Config> {
  const path = getConfigPath();
  try {
    const data = await readFile(path, 'utf-8');
    const parsed = JSON.parse(data) as Partial<Config>;
    const portfolio = Array.isArray(parsed.portfolio)
      ? parsed.portfolio.map((item) => ({
          market: String(item.market),
          code: String(item.code),
          name: String(item.name),
        }))
      : [];
    return {
      portfolio,
      refreshInterval: Math.max(3, Math.min(60, Number(parsed.refreshInterval) || 5)),
    };
  } catch (e: any) {
    if (e?.code === 'ENOENT') {
      const migrated = await migrateFromOld();
      if (migrated) {
        return { portfolio: migrated, refreshInterval: 5 };
      }
      return { portfolio: [], refreshInterval: 5 };
    }
    throw e;
  }
}

export async function save(config: Config): Promise<void> {
  const path = getConfigPath();
  const payload: Config = {
    portfolio: config.portfolio,
    refreshInterval: config.refreshInterval,
  };
  await writeFile(path, JSON.stringify(payload, null, 2) + '\n', 'utf-8');
}
